rfd/xxxx-command.md

```diff
@@ -0,0 +1,208 @@
+
+tsh ->> proxy: tsh command exec
+proxy ->> node1: execute command
+proxy ->> node2: execute command
+node1 ->> proxy: command output
+proxy ->> tsh: command output
+node2 ->> proxy: command output
+proxy ->> tsh: command output
+```
+Teleport Proxy will be used as the main command execution engine.
+
+* It will allow parallel execution of commands on multiple nodes (configurable execution pool size).
+* It will allow to stream output from multiple nodes to the client.
+* Nodes will record the session and upload in the same way as interactive session are recorded.
+* It will ensure that even in case of a network failure the command will continue to run on the node.
+  (all commands should be one shot. If an initial connection to the node fails, the proxy will retry.
+  If the connection is lost during the command execution, the command will continue to run on the node
+  and proxy will not re-try it as we have no guarantee that the command is idempotent).
+* (Optional) It will allow to specify a timeout for the command execution. If the command execution takes longer
+  than the specified timeout, the command will be terminated on the node and the output will be streamed
+  to the client.
+
+### Command resource
+
+We will introduce a new resource type `command` that will allow to define
+commands that can be executed by Teleport Command. The resource will have
+the following fields:
+
+```yaml
+#
+# Example resource for a Command
+#
+kind: command
+version: v1
+metadata:
+  # The name of the command. It must be unique.
+  name: cpu-usage
+  # Human-readable description
+  description: Show top 10 CPU usage
+spec:
+  # Interpreter to use for command execution. If not specified, /bin/sh will be used.
+  # This will allow to execute short Python or JS scripts.
+  interpreter: "/bin/bash"
+  command: |
+    ps -eo pid,ppid,cmd,%mem,%cpu --sort=-%cpu | head
+```
+
+Interpreter is an optional field. It should allow specifying the interpreter
+allowing to not only execute a shell command but also a short Python or JS script.
+
+#### Picking up the OS user
+
+The user will be provided as a command argument. If the user doesn't exist on the node,
+the command execution will fail. If a user want to execute a command as two different
+users, then one command should be execute for each user.
+An alternative would be to add the user to `ssh_service` configuration, but this
+will require modifying the configuration on all nodes. 
+
+## Roles
+
+Roles will be extended to allow defining access to commands. The following
+example will allow to execute commands with `environment=dev-cmd` label on nodes
+with `environment=dev` label. It will also deny access to commands with
+`environment=prod` label.
+This approach will allow to define a set of commands that can be executed
+on a particular set of nodes (e.g. `dev-cmd` commands can be executed on `dev` nodes
+but not on `prod` nodes).
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: developer
+spec:
```

Comment:
Will we use existing "logins" field to control Linux principals that users are allowed to run commands as?